### ISSUE TITLE
Help understanding 'error: failed retrieving file pkgname from disk : Couldn't open file' after aur sync

### ISSUE THREAD

**Issue by jwhendy** [AladW/aurutils#509]
Greetings, brand new user due to finding out `pacaur` was dead (how it took so long to learn this, I have no idea!). This is likely a noob question, so apologies in advance; if you can steer me in the right direction, I'd be happy to contribute to #217.

version: `aurutils-git 2.1.0.r0.g6c55b57-1`

Situation:
- I'm installing `ros-melodic-desktop-full`, a pretty large metapackage with some issues (bad dependencies, some downgrading required, etc.)
- when something fails during `aur sync1, I manually git clone from AUR and run `makepkg` to look more closely at the build files and debug
- now that I've switched to `aurutils` and a custom repository, my understanding was that I'd then add this to a local repository using `repo-add /var/cache/pacman/custom/custom.db.tar ./pkgname`
- when going back to the `aur sync` build, I'll get errors when pacman tries to install this package as a make depend. This package successfully built, can be installed with `pacman -U ./pkgfile`, and was added via `repo-add`

```
$ makepkg -s
$ repo-add /var/cache/pacman/custom/custom.db.tar ./ros-melodic-rostest-1.14.3-1-any.pkg.tar.xz
$ sudo pacman -Syu

$ sudo pacman -Ss ros-melodic-rostest
custom/ros-melodic-rostest 1.14.3-1
    ROS - Integration test suite based on roslaunch that is compatible with xUnit
    frameworks.
$ sudo pacman -S ros-melodic-rostest
resolving dependencies...
looking for conflicting packages...

Packages (1) ros-melodic-rostest-1.14.3-1

Total Installed Size:  0.10 MiB

:: Proceed with installation? [Y/n] 
(1/1) checking keys in keyring                          [#############################] 100%
(1/1) checking package integrity                        [#############################] 100%
:: File /var/cache/pacman/custom/ros-melodic-rostest-1.14.3-1-any.pkg.tar.xz is corrupted (invalid or corrupted package (checksum)).
Do you want to delete it? [Y/n] Y 
error: failed to commit transaction (invalid or corrupted package (checksum))
Errors occurred, no packages were upgraded.

### I chose to delete, and the error changes slightly afterward:
$ sudo pacman -Syyu
:: Synchronizing package databases...
 core                          135.0 KiB   643K/s 00:00 [#############################] 100%
 extra                        1692.9 KiB  4.78M/s 00:00 [#############################] 100%
 community                       4.8 MiB  7.46M/s 00:01 [#############################] 100%
 custom                        440.5 KiB  0.00B/s 00:00 [#############################] 100%
:: Starting full system upgrade...
 there is nothing to do
$ sudo pacman -Ss ros-melodic-rostest
custom/ros-melodic-rostest 1.14.3-1
    ROS - Integration test suite based on roslaunch that is compatible with xUnit
    frameworks.
$ sudo pacman -S ros-melodic-rostest
resolving dependencies...
looking for conflicting packages...

Packages (1) ros-melodic-rostest-1.14.3-1

Total Download Size:   0.03 MiB
Total Installed Size:  0.10 MiB

:: Proceed with installation? [Y/n] 
:: Retrieving packages...
error: failed retrieving file 'ros-melodic-rostest-1.14.3-1-any.pkg.tar.xz' from disk : Couldn't open file /var/cache/pacman/custom/ros-melodic-rostest-1.14.3-1-any.pkg.tar.xz
warning: failed to retrieve some files
error: failed to commit transaction (download library error)
Errors occurred, no packages were upgraded.
```

From the same AUR build directory:
```
$ sudo rm -r ./pkg ./src ./ros-melodic-rostest-1.14.3-*
$ ls
PKGBUILD
$ aur build
### it builds, then does it's magic on the db:
==> WARNING: An entry for 'ros-melodic-rostest-1.14.3-1' already existed
  -> Computing checksums...
  -> Removing existing entry 'ros-melodic-rostest-1.14.3-1'...
  -> Creating 'desc' db entry...
  -> Creating 'files' db entry...
==> Creating updated database file '/var/cache/pacman/custom/custom.db.tar'
:: Synchronizing package databases...
 custom                          2.6 MiB   878M/s 00:00 [#############################] 100%
:: Synchronizing package databases...
 custom                        440.5 KiB   108M/s 00:00 [#############################] 100%
:: Starting full system upgrade...
 there is nothing to do

### no issues on the install
$ sudo pacman -S ros-melodic-rostest
resolving dependencies...
looking for conflicting packages...

Packages (1) ros-melodic-rostest-1.14.3-1

Total Installed Size:  0.10 MiB

:: Proceed with installation? [Y/n] 
(1/1) checking keys in keyring                          [#############################] 100%
(1/1) checking package integrity                        [#############################] 100%
(1/1) loading package files                             [#############################] 100%
(1/1) checking for file conflicts                       [#############################] 100%
(1/1) checking available disk space                     [#############################] 100%
:: Processing package changes...
(1/1) installing ros-melodic-rostest                    [#############################] 100%
```

From various other comments, I'm guessing this is because I'm not doing things properly, but I'm not entirely sure why.

I'm sure this is an elementary noob error, but am not sure what's happening.

**Comment by AladW** [comment]
The issue here is that you have a built package from a previous build in
the pacman cache (in /var/cache/pacman/custom), but have a package of
the same name (and pkgver) from a new build (in ./) in your local
repository. As both packages will have differing checksums, pacman will
complain. See issue #85 for details.

Until now, the suggested approach was to keep all packages in a pacman
CacheDir, and use aur-build - which will move all packages to that
CacheDir, and run repo-add with the correct path. This however means
that accidentally running pacman -Scc would remove all your local
repository packages, so I'm rewriting aur(1) to use (read-only) bind
mounts instead.

Note that neither approach is mandatory - it just avoids the problematic
which happens when rebuilding a package without bumping its pkgver.

**Comment by jwhendy** [comment]
I sort of follow, but am not sure on terminology and what is actually conflicting. As in, I thought `/var/cache/pacman/custom` *was* my "local repository" which you're calling `./`. And my assumption was that `repo-add` put this build package into `/var/cache/pacman/custom`. If `aur build` also uses `makepkg`, is there a reason the two method result in different checksums?

If they shouldn't, it must be due to mucking with dependencies and having the package build against different libraries, perhaps.

Sorry to be dense and I appreciate the input/coaching!

**Comment by AladW** [comment]
Maybe I misread your original report. Were you running repo-add from
/var/cache/pacman/custom, or from the directory that contained a
PKGBUILD?

Note that repo-add itself does not move packages - it only adds the
package name (%FILENAME%) and package checksums (%SHA256SUM%, %MD5SUM%,
...) to the database, that is, to the .db file in /var/cache/pacman/custom.

For example, assume you have /var/cache/pacman/custom/foo-1.pkg.tar.xz
with md5sum b026324c6904b2a9cb4b88d6d61c81d1 and $HOME/foo-1.pkg.tar.xz
with md5sum 26ab0db90d72e28ad0ba1e22ee510510. Adding either with
repo-add will result in

%FILENAME%
foo-1.pkg.tar.xz
...
%MD5SUM%
b026324c6904b2a9cb4b88d6d61c81d1

and

%FILENAME%
foo-1.pkg.tar.xz
...
%MD5SUM%
26ab0db90d72e28ad0ba1e22ee510510.

entries being added to /var/cache/pacman/custom/custom.db, respectively.
Now, if the /var/cache/pacman/custom directory contains the file with
the former md5sum, but the /var/cache/pacman/custom/custom.db file
contains the md5sum to the package in your home directory, there will be
a mismatch:

File /var/cache/pacman/custom/foo-1-x86_64.pkg.tar.xz is corrupted (invalid or corrupted package (checksum)).

As to why these packages may differ between several builds, well there
are many factors, even apparently trivial ones such as a beginning
starting date. It will still take a while before we have reproducible
builds [1] everywhere.

[1] https://wiki.debian.org/ReproducibleBuilds

**Comment by jwhendy** [comment]
So immensely helpful. Indeed, I was running `repo-add` from the build directory I use at `~/installed/build/foo`. I did not know it didn't move the package. I wonder if this could be in some FAQ somewhere as `man aur` points folks to look at [the arch wiki](), and that has this line:

>Storing all the built packages to be included in the repository in one directory also allows to use shell glob expansion to add or update multiple packages at once:
```
 repo-add /path/to/repo.db.tar.gz /path/to/*.pkg.tar.xz
```

To be fair, it also says:

>The database and the packages do not need to be in the same directory when using repo-add, but keep in mind that when using pacman with that database, they should be together. 

So, that second part now makes more sense, but being a noob to this, when I first read it I simply thought "Oh, cool. Now that I'm using this local repo thing, I can add all the packages I've already built with `makepkg` in ~/installed/build` to my shiny new db." To do this over again, I'd move them to `/var/cache/pacman/custom` first, then add (at least this sounds like the way to get them on the `aur sync/build` radar).

It seems like this was my issue after all and either the build genuinely changed, or to you point something more nuanced changed. In any case, I think my question is answered!
